### DISCUSSION TITLE
Is L2Norm better than cosine similarity?

### DISCUSSION THREAD

**Discussion by syeds-git** [shubham0204/FaceRecognition_With_FaceNet_Android#10] [Q&A]
I used similar approach in my flutter app but did not get satisfactory results using L2Norm. I then switched to cosine and started to get reasonable results. For L2Norm I am getting results which are somehow more than 1 usually between 10 & 20 and can't be used to identify the face. However using cosine the results are much more clearer. What sort of number you expect to get out of L2Norm? should they be between 0 & 1?

**Comment by shubham0204** [comment]
We can understand the difference from here. L2 Norm or the Euclidean Norm is a function which takes in a vector and returns a non-negative real number. In our case, we are calculating the Euclidean norm of the vector `user_embedding - subject_embedding`. The magnitude of this vector actually gives us the euclidean distance between the vectors `user_embedding` and `subject_embedding` in hyperspace. Calculating the L2 norm of the `user_embedding - subject_embedding` is nothing but calculating the euclidean distance between these two vectors in hyperspace.
Now, cosine similarity returns the cosine of the angle between these vectors, The range of cosine function is `[ -1 , 1 ]`, so the cosine similarity also holds this range. A score of 1 would mean that both the vectors are parallel ( as `cos( 0 ) = 1` ) whereas a score of -1 would mean the vectors are anti-parallel ( as `cos( 180 ) = -1` ).
Two vectors might be far away from our `subject_embedding` but may have the same angle with `subject_embedding`. Thus cosine similarity would provide the same score for both these vectors.
As of my research, L2 norm should perform better than cosine similarity but we can keep both of them in the project.
